specification/hugr.md

```diff
@@ -1168,44 +1168,78 @@ The new hugr is then derived as follows:
+  - it has no root node (the set of nodes without a parent is
+    denoted $\mathrm{top}(G)$ );
+  - it may have empty container nodes (the set of these is
+    denoted $\mathrm{bot}(G)$ ).
 
-  - a map B from bot(G) to IDs of container nodes in S\*, such that B(x)
-    is separated from B(y) unless x == y. Let X be the set of children
-    of values in B, and R be S\*\\X\*.
+A partial hugr describes a set of nodes and well-formed edges between
+them that potentially forms a subgraph of a hugr.
 
-  - a bijection μ<sub>inp</sub> between inp(G) and the set of input
-    ports of nodes in R whose source is not in R;
+Given a set $S$ of nodes in a hugr, let $S^\*$ be the set of all nodes
+descended from nodes in $S$ (i.e. reachable from $S$ by following hierarchy edges),
+including $S$ itself.
 
-  - a bijection μ<sub>out</sub> between out(G) and the set of output
-    ports of nodes in R whose target is not in R;
+Call two nodes $a, b \in \Gamma$ _separated_ if $a \notin \\{b\\}^\*$ and
+$b \notin \\{a\\}^\*$ (i.e. there is no hierarchy relation between them).
 
-  - disjoint subsets Init and Term of top(G);
+We define two structures to specify edge insertion.
 
-The new hugr is then derived by:
+A `NewInpEdgeSpec` specifies an edge inserted from an existing node to a (given) new node.
+It contains the following fields:
 
-1.  adding the new nodes from G;
+  - `EdgeKind`: may be `Value`, `Order`, `Static` or `ControlFlow`.
+  - `SrcNode`: the source node of the new edge.
+  - `SrcPort`: (for `Value` and `Static` edges only) the source port.
+  - `ToPos`: (for `Value` and `Static` edges only) the desired position among
+    the incoming ports to the new node.
 
-2.  connecting the ports according to the bijections μ<sub>inp</sub> and
-    μ<sub>out</sub>;
+A `NewOutEdgeSpec` specifies an edge inserted from a new node to an existing node.
+It contains the following fields:
 
-3.  for each node n in top(G), adding a hierarchy edge from t(n) to n,
-    placing n in the first position among children of t(n) if n is in
-    Init and in the second position if n is in Term;
+  - `EdgeKind`: may be `Value`, `Order`, `Static` or `ControlFlow`.
+  - `TgtNode`: the target node of the new edge.
+  - `TgtPort`: (for `Value` and `Static` edges only) the target port (whose
+    existing incoming edge will be removed).
+  - `FromPos` (for `Value`, `Static` and `ControlFlow` edges only) the desired
+    position among the outgoing ports from the new node.
 
-4.  for each node n in bot(G), and for each child m of b(n), adding a
-    hierarchy edge from n to m (replacing m’s existing parent edge)
+The `Replace` method takes as input:
 
-5.  removing all nodes in R
+  - a set $S$ of mutually-separated nodes in $\Gamma$;
+  - a partial hugr $G$;
+  - a map $T : \mathrm{top}(G) \to \Gamma \setminus S^*$ whose image consists of container nodes;
+  - a map $B : \mathrm{bot}(G) \to S^\*$ whose image consists of container nodes, such that $B(x)$
```

Comment:
Note that AFAICS `bot` and `top` are defined in the *first* version of partial hugr, [here](https://github.com/CQCL-DEV/hugr/blob/96cac0e9064d7ec3ee98a351330bc4577c05dde2/specification/hugr.md?plain=1#L1111)